src/psbt.c

```diff
@@ -0,0 +1,2302 @@
+    unsigned char *final_script_sig,
+    size_t final_script_sig_len,
+    struct wally_tx_witness_stack *final_witness,
+    struct keypath_map *keypaths,
+    struct partial_sigs_map *partial_sigs,
+    struct unknowns_map *unknowns,
+    uint32_t sighash_type,
+    struct wally_psbt_input **output)
+{
+    struct wally_psbt_input *result;
+    int ret = WALLY_OK;
+
+    TX_CHECK_OUTPUT;
+    TX_OUTPUT_ALLOC(struct wally_psbt_input);
+
+    ret = wally_psbt_input_update(result,
+                                non_witness_utxo,
+                                witness_utxo,
+                                redeem_script,
+                                redeem_script_len,
+                                witness_script,
+                                witness_script_len,
+                                final_script_sig,
+                                final_script_sig_len,
+                                final_witness,
+                                keypaths,
+                                partial_sigs,
+                                unknowns,
+                                sighash_type);
+
+    if (ret != WALLY_OK) {
+        clear_and_free(result, sizeof(*result));
+        *output = NULL;
+    }
+    return ret;
+}
+
+int wally_psbt_input_update(
+    struct wally_psbt_input *input,
+    struct wally_tx *non_witness_utxo,
+    struct wally_tx_output *witness_utxo,
+    unsigned char *redeem_script,
+    size_t redeem_script_len,
+    unsigned char *witness_script,
+    size_t witness_script_len,
+    unsigned char *final_script_sig,
+    size_t final_script_sig_len,
+    struct wally_tx_witness_stack *final_witness,
+    struct keypath_map *keypaths,
+    struct partial_sigs_map *partial_sigs,
+    struct unknowns_map *unknowns,
+    uint32_t sighash_type)
+{
+    int ret = WALLY_OK;
+
+    if ((non_witness_utxo && clone_tx(non_witness_utxo, &input->non_witness_utxo) != WALLY_OK) ||
+        (witness_utxo && wally_tx_output_init_alloc(witness_utxo->satoshi, witness_utxo->script, witness_utxo->script_len, &input->witness_utxo) != WALLY_OK) ||
+        (final_witness && !(input->final_witness = clone_witness(final_witness))) ||
+        !clone_bytes(&input->redeem_script, redeem_script, redeem_script_len) ||
+        !clone_bytes(&input->witness_script, witness_script, witness_script_len) ||
+        !clone_bytes(&input->final_script_sig, final_script_sig, final_script_sig_len) ||
+        (keypaths && !(input->keypaths = clone_keypath_map(keypaths))) ||
+        (partial_sigs && !(input->partial_sigs = clone_partial_sigs_map(partial_sigs))) ||
+        (unknowns && !(input->unknowns = clone_unknowns_map(unknowns))))
+        ret = WALLY_ENOMEM;
+    if (ret == WALLY_OK) {
+        input->redeem_script_len = redeem_script_len;
+        input->witness_script_len = witness_script_len;
+        input->final_script_sig_len = final_script_sig_len;
+        input->sighash_type = sighash_type;
+    }
+
+    return ret;
```

Comment:
this function can potentially leak memory.

Reply:
I believe I've fixed that. Also fixed `wally_psbt_output_update` which had the same potential leaks.